File: src/ai/extraction/extraction.prompt.ts
import { AgentSession } from '../shared/agent.types'

export function buildExtractionPrompt(session: AgentSession, availableSkills: string[]): string {
  return `
You are a senior software architect and project analyst for SkillBridge freelance platform.

Analyze the conversation below and extract structured project information.

CONVERSATION HISTORY:
${session.history.map(m => `${m.role.toUpperCase()}: ${m.content}`).join('\n\n')}

AVAILABLE SKILLS IN DATABASE:
${availableSkills.join(', ')}

INSTRUCTIONS:
1. Extract ALL project information mentioned anywhere in the conversation.
2. If a field is not mentioned, set it to null.
3. For features, extract every feature mentioned as an array of strings.
4. For budget, extract min and max as numbers in USD. If only one number mentioned, use it for both.
5. For timeline, extract as a human readable string e.g. "2 months", "8 weeks".
6. For techPreferences - ANALYZE the project deeply as a senior architect:
   - Look at projectType, platform, features, and complexity.
   - Pick the BEST matching technologies from the "AVAILABLE SKILLS IN DATABASE" list above.
   - If a core technology is not in the list but absolutely required, you may still include it.
   - NEVER leave techPreferences empty — always infer from project context.
7. For expertiseNeeded, infer from project complexity: "entry", "intermediate", or "senior".
8. For platform, extract e.g. "iOS", "Android", "iOS + Android", "Web", "Cross-platform".
9. For projectType, extract e.g. "Mobile App", "Web App", "REST API", "E-commerce", etc.

RETURN STRICT JSON ONLY. NO markdown. NO explanation. NO extra text.

{
  "projectType": string | null,
  "platform": string | null,
  "features": string[],
  "budgetMin": number | null,
  "budgetMax": number | null,
  "timeline": string | null,
  "techPreferences": string[],
  "expertiseNeeded": "entry" | "intermediate" | "senior" | null,
  "additionalNotes": string | null
}
`.trim()
}

export function buildExtractionCheckPrompt(session: AgentSession): string {
  return `
You are checking if enough project information has been collected to proceed to freelancer matching.

CURRENT EXTRACTED PROJECT DATA:
${JSON.stringify(session.project, null, 2)}

CONVERSATION HISTORY:
${session.history.map(m => `${m.role.toUpperCase()}: ${m.content}`).join('\n\n')}

REQUIRED FIELDS TO PROCEED:
- projectType (what kind of project)
- features (at least 1 feature)
- budgetMin and budgetMax (budget range)
- timeline (deadline or duration)

RETURN STRICT JSON ONLY:
{
  "isComplete": true | false,
  "missingFields": string[],
  "confidence": number
}
`.trim()
}

File: src/ai/extraction/extraction_prompt.py
from typing import Dict, Any, List


def build_extraction_prompt(session: Dict[str, Any], available_skills: List[str]) -> str:
    history_text = "\n\n".join(
        f"{m['role'].upper()}: {m['content']}"
        for m in session.get("history", [])
    )
    skills_text = ", ".join(available_skills)

    return f"""
ROLE: Senior Software Architect & Project Analyst.
MISSION: Extract the 5 REQUIRED PROJECT FIELDS from the conversation history.

CONVERSATION HISTORY:
{history_text}

REFERENCE SKILLS (Use these for 'techPreferences' if they match):
{skills_text}

EXTRACTION RULES:
1. PROJECT GOAL: What is the core idea? (Extract into 'projectType')
2. FEATURES: List of specific functionalities mentioned.
3. BUDGET: Extract number or range. If a single number, set budgetMin and budgetMax to that value.
4. TIMELINE: Extract duration or deadline.
5. PLATFORM: Web, iOS, Android, Desktop, AI System, etc.

ADDITIONAL INFERENCE:
- techPreferences: Based on the Platform and Features, pick the BEST stack. prioritize the REFERENCE SKILLS list.
- expertiseNeeded: "entry", "intermediate", or "senior" based on technical complexity.

RETURN STRICT JSON ONLY:
{{
  "projectType": string | null,
  "platform": string | null,
  "features": string[],
  "budgetMin": number | null,
  "budgetMax": number | null,
  "timeline": string | null,
  "techPreferences": string[],
  "expertiseNeeded": "entry" | "intermediate" | "senior" | null,
  "additionalNotes": string | null
}}
""".strip()


def build_extraction_check_prompt(session: Dict[str, Any]) -> str:
    import json
    project_json = json.dumps(session.get("project", {}), indent=2)

    return f"""
DECISION ENGINE: Is this project ready for matching?

CURRENT PROJECT DATA:
{project_json}

HARD REQUIRMENTS (ALL MUST BE PRESENT):
1. Project Idea/Goal (projectType)
2. Features (features list cannot be empty)
3. Budget (budgetMin or budgetMax)
4. Timeline (timeline)
5. Platform (platform)
6. Tech Stack (techPreferences list cannot be empty)

PROTOCOL: If ANY field is missing, 'isComplete' MUST be FALSE.

RETURN STRICT JSON ONLY:
{{
  "isComplete": boolean,
  "missingFields": string[],
  "confidence": number (0-100)
}}
""".strip()